Unit test failure in Jenkins

**Problem description**
`io.pravega.segmentstore.server.reading.ContainerReadIndexTests.testCacheEviction` failed with
log message :
`java.lang.AssertionError: Unexpected CacheKey.SegmentOffset at index 11 for SegmentId 79. expected:<3600> but was:<3700>`

Stack trace:
```
java.lang.AssertionError: Unexpected CacheKey.SegmentOffset at index 11 for SegmentId 79. expected:<3600> but was:<3700>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:834)
	at org.junit.Assert.assertEquals(Assert.java:645)
	at io.pravega.segmentstore.server.reading.ContainerReadIndexTests.checkOffsets(ContainerReadIndexTests.java:1042)
	at io.pravega.segmentstore.server.reading.ContainerReadIndexTests.testCacheEviction(ContainerReadIndexTests.java:879)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:298)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:292)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.lang.Thread.run(Thread.java:745)
```
**Problem location**


**Suggestions for an improvement**

--- a/segmentstore/server/src/test/java/io/pravega/segmentstore/server/reading/ContainerReadIndexTests.java
+++ b/segmentstore/server/src/test/java/io/pravega/segmentstore/server/reading/ContainerReadIndexTests.java
@@ -47,6 +47,7 @@ import java.util.List;
 import java.util.Map;
 import java.util.Random;
 import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.CompletionException;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicInteger;
@@ -809,10 +810,15 @@ public class ContainerReadIndexTests extends ThreadPooledTestSuite {
                 ReadResult result = context.readIndex.read(segmentId, offset, appendSize, TIMEOUT);
                 ReadResultEntry resultEntry = result.next();
                 Assert.assertEquals("Unexpected type of ReadResultEntry when trying to load up data into the ReadIndex Cache.", ReadResultEntryType.Storage, resultEntry.getType());
+                CompletableFuture<Void> insertedInCache = new CompletableFuture<>();
+                context.cacheFactory.cache.insertCallback = ignored -> insertedInCache.complete(null);
                 resultEntry.requestContent(TIMEOUT);
                 ReadResultEntryContents contents = resultEntry.getContent().get(TIMEOUT.toMillis(), TimeUnit.MILLISECONDS);
                 Assert.assertFalse("Not expecting more data to be available for reading.", result.hasNext());
                 Assert.assertEquals("Unexpected ReadResultEntry length when trying to load up data into the ReadIndex Cache.", appendSize, contents.getLength());
+
+                // Wait for the entry to be inserted into the cache before moving on.
+                insertedInCache.get(TIMEOUT.toMillis(), TimeUnit.MILLISECONDS);
             }
 
             context.cacheManager.applyCachePolicy();
@@ -1347,6 +1353,7 @@ public class ContainerReadIndexTests extends ThreadPooledTestSuite {
     //region TestCache
 
     private static class TestCache extends InMemoryCache {
+        Consumer<CacheKey> insertCallback;
         Consumer<CacheKey> removeCallback;
 
         TestCache(String id) {
@@ -1354,13 +1361,21 @@ public class ContainerReadIndexTests extends ThreadPooledTestSuite {
         }
 
         @Override
+        public void insert(Cache.Key key, byte[] payload) {
+            super.insert(key, payload);
+            Consumer<CacheKey> callback = this.insertCallback;
+            if (callback != null) {
+                callback.accept((CacheKey) key);
+            }
+        }
+
+        @Override
         public void remove(Cache.Key key) {
+            super.remove(key);
             Consumer<CacheKey> callback = this.removeCallback;
             if (callback != null) {
                 callback.accept((CacheKey) key);
             }
-
-            super.remove(key);
         }
     }
 

